### PR TITLE
refactor(types)!: use user-defined types with Cosmos DB

### PR DESCRIPTION
**Pull request by dciborow**
## Description

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [ ] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ ] I have updated the examples in README with the latest module version number.

### CODE REVIEW

File: modules/storage/cosmos-db/module/cosmos.bicep [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  properties: _properties
+}
+
+@description('The resource name.')
+@minLength(3)
+@maxLength(64)
+type _name = string
+
+@description('The location of the resource group to which the resource belongs.')
+type _location = 'eastus' | 'eastus2' | 'westus' | 'westus2' | 'westcentralus' | 'westeurope' | 'southeastasia' | 'northeurope' | 'uksouth' | 'ukwest' | 'australiaeast' | 'australiasoutheast' | 'brazilsouth' | 'southindia' | 'centralindia' | 'westindia' | 'canadacentral' | 'canadaeast' | 'japaneast' | 'japanwest' | 'koreacentral' | 'koreasouth' | 'francecentral' | 'southafricanorth' | 'uaenorth' | 'switzerlandnorth' | 'germanywestcentral' | 'norwayeast'
```
Comment (shenglol) [inline modules/storage/cosmos-db/module/cosmos.bicep:31]: We cannot make location a type. The available locations depend on the Cloud environments the module is going to be deployed to and the SKU of the resources for some resource types. Besides, if ARM spins up a new location, all existing versions of the module will be broken.

File: modules/storage/cosmos-db/module/cosmos.bicep [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+@maxLength(128)
+type _tag_key = string
+
+@maxLength(256)
+type _tag_value = string
+
+@description('Tags are a list of key-value pairs that describe the resource. These tags can be used in viewing and grouping this resource (across resource groups). A maximum of 15 tags can be provided for a resource. Each tag must have a key no greater than 128 characters and value no greater than 256 characters. For example, the default experience for a template type is set with "defaultExperience": "Cassandra". Current "defaultExperience" values also include "Table", "Graph", "DocumentDB", and "MongoDB".')
+@maxLength(15)
+type _tags = {
```
Comment (shenglol) [inline modules/storage/cosmos-db/module/cosmos.bicep:41]: I think we should set up a naming convention for types (and document that). Personally, I don't like mixing up camel case and snake case. How about we follow TypeScript's convention to use pascal case for types?

Comment (dciborow) [reply modules/storage/cosmos-db/module/cosmos.bicep:41]: sounds good. Let me add notes into the contribution guide with this PR to include these conventions.

File: modules/storage/cosmos-db/module/cosmos.bicep (hunk 1 above)
Comment (dciborow) [reply modules/storage/cosmos-db/module/cosmos.bicep:31]: ill roll this out. Need to think about how to make location easier for the user to select, without also restricting their options.

**Pull request closed without merging**